builtin/logical/transit/path_decrypt.go

```diff
@@ -84,24 +153,43 @@ func (b *backend) pathDecryptWrite(
 		return logical.ErrorResponse("policy not found"), logical.ErrInvalidRequest
 	}
 
-	plaintext, err := p.Decrypt(context, nonce, ciphertext)
-	if err != nil {
-		switch err.(type) {
-		case errutil.UserError:
-			return logical.ErrorResponse(err.Error()), logical.ErrInvalidRequest
-		case errutil.InternalError:
-			return nil, err
-		default:
-			return nil, err
+	for i, item := range batchInputItems {
+		if batchResponseItems[i].Error != "" {
+			continue
+		}
+
+		plaintext, err := p.Decrypt(item.DecodedContext, item.DecodedNonce, item.Ciphertext)
+		if err != nil {
+			switch err.(type) {
+			case errutil.UserError:
+				return logical.ErrorResponse(err.Error()), logical.ErrInvalidRequest
+			case errutil.InternalError:
+				return nil, err
+			default:
+				return nil, err
+			}
 		}
+		batchResponseItems[i].Plaintext = plaintext
 	}
 
-	// Generate the response
-	resp := &logical.Response{
-		Data: map[string]interface{}{
-			"plaintext": plaintext,
-		},
+	resp := &logical.Response{}
+	if len(batchInputRaw) != 0 {
+		batchResponseJSON, err := jsonutil.EncodeJSON(batchResponseItems)
+		if err != nil {
+			return nil, fmt.Errorf("failed to JSON encode batch response")
+		}
+		resp.Data = map[string]interface{}{
+			"data": string(batchResponseJSON),
+		}
+	} else {
+		if batchResponseItems[0].Error != "" {
```

Comment:
Given that all of the cases where this would be true are user/input errors, this should probably be a logical.ErrorResponse call, not an internal error message.